pkg/envoy/xds/watcher.go

```diff
@@ -118,6 +118,17 @@ func (w *ResourceWatcher) WatchResources(ctx context.Context, typeURL string, la
 
 	for ctx.Err() == nil && res == nil {
 		w.versionLocker.Lock()
+		// If the client ACKed a version that we have never sent back, this
+		// indicates that this server restarted but the client survived and had
+		// received a higher version number from the previous server instance.
+		// Bump the resource set's version number to match the client's and
+		// send a response immediately.
+		if waitForVersion && w.version < waitVersion {
+			w.versionLocker.Unlock()
+			w.resourceSet.EnsureVersion(typeURL, waitVersion+1)
+			w.versionLocker.Lock()
+		}
+
 		for ctx.Err() == nil && waitForVersion && w.version <= waitVersion {
```

Comment:
Should the changes made by `EnsureVersion()` have an effect on these conditions? `EnsureVersion()` changes the version number of the cache (`w.resourceSet`), but no the local variables used in the condition here.

Reply:
Actually, it does have an effect on `w.version` because the resource set will call back to w to update its version number. That's why I have to temporarily unlock `w.versionLocker`, to let the callback execute and avoid a deadlock.

I will add a comment.

Reply:
Added this comment:
```
			// Calling EnsureVersion will increase the version of the resource
			// set, which in turn will callback w.HandleNewResourceVersion with
			// that new version number. In order for that callback to not
			// deadlock, temporarily unlock w.versionLocker.
			// The w.HandleNewResourceVersion callback will update w.version to
			// the new resource set version.
```